src/proof/theory_proof.cpp

```diff
@@ -1264,4 +1263,387 @@ void TheoryProof::printRewriteProof(std::ostream& os, const Node &n1, const Node
+      // correctly.
+      std::vector<std::shared_ptr<theory::eq::EqProof>> orderedEqualities;
+
+      // Insert target clause first.
+      if (targetAppearsBefore)
+      {
+        orderedEqualities.push_back(pf.d_children[i - 1]);
+        // The target has already been added to subTrans; remove it.
+        subTrans->d_children.pop_back();
+      }
+      else
+      {
+        orderedEqualities.push_back(pf.d_children[i + count]);
+      }
+
+      // Start with the congruence closure closest to the target clause, and
+      // work our way back/forward.
+      if (targetAppearsBefore)
+      {
+        for (unsigned j = 0; j < count; ++j)
+        {
+          if (pf.d_children[i + j]->d_children[0]->d_id
+              != theory::eq::MERGED_THROUGH_REFLEXIVITY)
+            orderedEqualities.insert(orderedEqualities.begin(),
+                                     pf.d_children[i + j]->d_children[0]);
+          if (pf.d_children[i + j]->d_children[1]->d_id
+              != theory::eq::MERGED_THROUGH_REFLEXIVITY)
+            orderedEqualities.insert(orderedEqualities.end(),
+                                     pf.d_children[i + j]->d_children[1]);
+        }
+      }
+      else
+      {
+        for (unsigned j = 0; j < count; ++j)
+        {
+          if (pf.d_children[i + count - 1 - j]->d_children[0]->d_id
+              != theory::eq::MERGED_THROUGH_REFLEXIVITY)
+            orderedEqualities.insert(
+                orderedEqualities.begin(),
+                pf.d_children[i + count - 1 - j]->d_children[0]);
+          if (pf.d_children[i + count - 1 - j]->d_children[1]->d_id
+              != theory::eq::MERGED_THROUGH_REFLEXIVITY)
+            orderedEqualities.insert(
+                orderedEqualities.end(),
+                pf.d_children[i + count - 1 - j]->d_children[1]);
+        }
+      }
+
+      // Copy the result into the main transitivity proof.
+      subTrans->d_children.insert(subTrans->d_children.end(),
+                                  orderedEqualities.begin(),
+                                  orderedEqualities.end());
+
+      // Increase i to skip over the children that have been processed.
+      i += count;
+      if (targetAppearsAfter)
+      {
+        ++i;
+      }
+    }
+
+    // Else, just copy the child proof as is
+    else
+    {
+      subTrans->d_children.push_back(pf.d_children[i]);
+      ++i;
+    }
+  }
+
+  bool disequalityFound = (*neg >= 0);
+  if (!disequalityFound)
+  {
+    Debug("pf::" + theoryName)
+        << "A disequality was NOT found. UNSAT due to merged constants"
+        << std::endl;
+    Debug("pf::" + theoryName) << "Proof for: " << pf.d_node << std::endl;
+    Assert(pf.d_node.getKind() == kind::EQUAL);
+    Assert(pf.d_node.getNumChildren() == 2);
+    Assert(pf.d_node[0].isConst() && pf.d_node[1].isConst());
+  }
+}
+
+void TheoryProof::identicalEqualitiesPrinterHelper(theory::TheoryId theoryId,
```

Comment:
Maybe `TheoryProof` should have a virtual method `getTheoryId`, which can be used to get the `theoryId` instead of passing it in here.